sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Strings;
+import java.net.URI;
+
+/**
+ * Utility class for handling file paths.
+ */
+public class PathUtils {
+
+  private static final String URI_DELIMITER = "/";
+
+  /**
+   * Resolve multiple {@code others} against the {@code directory} sequentially.
+   *
+   * <p>Unlike {@link URI#resolve}, {@link #resolveAgainstDirectory} includes the last segment of
+   * the path.
+   *
+   * <p>Empty paths in {@code others} are ignored. If {@code others} contains one or more
+   * absolute paths, then this method returns a path that starts with the last absolute path
+   * in {@code others} joined with the remaining paths.
+   */
+  public static String resolveAgainstDirectory(String directory, String... others) {
+    String ret = directory;
+    for (String other : others) {
+      ret = resolveAgainstDirectory(ret, other);
+    }
+    return ret;
+  }
+
+  private static String resolveAgainstDirectory(String directory, String other) {
+    if (Strings.isNullOrEmpty(other)) {
+      return directory;
+    }
+    URI dirUri;
+    if (directory.endsWith(URI_DELIMITER)) {
+      dirUri = URI.create(directory);
+    } else {
+      dirUri = URI.create(directory + URI_DELIMITER);
+    }
+    return dirUri.resolve(other).toString();
+  }
+
+  /**
+   * Returns the directory portion of the {@code path}, which includes all but the file name.
```

Comment:
Document what does this return if the path ends with a / ?
What does this return if this is a path something like "gs://bucket"?
What does this return for ".." or "." or "~" ? Are these even allowed?
etc.
In general, examples would be very helpful in docs to the methods of this whole class. There was a lot of ambiguity in docs of IOChannelUtils and since we're revamping the whole thing it'd be great to also do comments properly.